src/v2.ts

```diff
@@ -0,0 +1,197 @@
+  cloudEventOverride?: CloudEventOverrides): CloudEvent => {
+  const type = cloudEventOverride?.type || _getCloudEventType(cloudFunction) || '';
+  return {
+    ..._createCloudEventWithDefaultValues(cloudFunction, type),
+    ...cloudEventOverride,
+    type
+  };
+};
+
+/** @internal */
+
+const _getCloudEventSource =
+  (cloudFunction: CloudFunction<any>, type: string): string => {
+  const projectId = '__PROJECT_ID__';
+  switch (type) {
+      case 'google.cloud.storage.object.v1.archived': // Fall-through intended
+      case 'google.cloud.storage.object.v1.deleted': // Fall-through intended
+      case 'google.cloud.storage.object.v1.finalized': // Fall-through intended
+      case 'google.cloud.storage.object.v1.metadataUpdated':
+        return `//storage.googleapis.com/projects/_/buckets/${projectId}`;
+      case 'google.cloud.pubsub.topic.v1.messagePublished':
+        const topicId = cloudFunction?.__endpoint?.eventTrigger?.eventFilters?.topic || '';
+        return `//pubsub.googleapis.com/projects/${projectId}/topics/${topicId}`;
+        return cloudFunction?.__endpoint?.eventTrigger?.eventFilters?.topic || '';
+      case 'google.firebase.firebasealerts.alerts.v1.published':
+        return `//firebasealerts.googleapis.com/projects/${projectId}`;
+      default:
+        return '';
+    }
+  };
+
+const _getCloudEventSubject =
+  (cloudFunction: CloudFunction<any>, type: string): string => {
+    switch (type) {
+      case 'google.cloud.storage.object.v1.archived': // Fall-through intended
+      case 'google.cloud.storage.object.v1.deleted': // Fall-through intended
+      case 'google.cloud.storage.object.v1.finalized': // Fall-through intended
+      case 'google.cloud.storage.object.v1.metadataUpdated':
+        return `objects/__STORAGE_FILENAME__`;
+      case 'google.firebase.firebasealerts.alerts.v1.published': // FirebaseAlerts do not contain a subject
+      case 'google.cloud.pubsub.topic.v1.messagePublished': // Pubsub CloudEvents do not contain a subject
+      default:
+        return null;
+    }
+  };
+
+const _getCloudEventType =
+  (cloudFunction: CloudFunction<any>): string => {
+    return cloudFunction?.__endpoint?.eventTrigger?.eventType || '';
+  };
+
+/** @return CloudEvent populated with default values */
+export const _createCloudEventWithDefaultValues =
+  <T>(cloudFunction: CloudFunction<T>, type: string): CloudEvent => {
+  const event = {
+    id: _makeEventId(),
+    source: _getCloudEventSource(cloudFunction, type),
+    type,
+    data: {},
+    time: new Date().toISOString(),
+    params: {}
+  } as CloudEvent;
+
+  const subject = _getCloudEventSubject(cloudFunction, type);
+  if (subject) { event.subject = subject; }
+
+  return (event);
+};
+
+function _makeEventId(): string {
+  return (
+    Math.random()
+      .toString(36)
+      .substring(2, 15) +
+    Math.random()
+      .toString(36)
+      .substring(2, 15)
+  );
+}
+
+function _checkCloudEventValidity(
```

Comment:
CloudEvents allow arbitrary extensions, which just show up as extra fields in the JSON format. This is well intentioned, but unnecessary.